Add get!T method to DBusAny

thaven opened the pull request:
Efficiently get values without type conversion.

This is also meant to somewhat align `DBusAny` with Phobos variants.

A follow-up PR will refactor `DBusAny.to`, using `get` where possible.

BTW, I know the unittest is ugly. It will be better after the refactor of `to`, because then `get` can be tested by just testing `to`.

thaven committed:
Fix typo in DBusAny.toString

thaven committed:
Add ddbus.util.BasicTypes

thaven committed:
Implement DBusAny.get

thaven committed:
Add unittest for DBusAny.get

WebFreak001 commented on source/ddbus/exception.d:
```diff
@@ -36,7 +36,9 @@ class DBusException : Exception {
+  of its actual value.
 +/
 class TypeMismatchException : Exception {
   package this(
```
Maybe we should actually make these public so they can be reused in user code?

thaven replied:
It doesn't make much sense to make this one public, as it produces quite a specific message. I changed te other overload (the one with a `message` parameter) to be public.

thaven committed:
Make TypeMismatchException constructor public

WebFreak001 commented on source/ddbus/thin.d:
```diff
@@ -264,6 +267,66 @@ struct DBusAny {
   }
 
+  /++
+    Returns the value stored in the DBusAny object by specifying the type.
```
you should change this to `Returns: ...` because ddoc supports that as special syntax

WebFreak001 commented on source/ddbus/thin.d:
```diff
@@ -264,6 +267,66 @@ struct DBusAny {
+      return obj;
+    } else static if (is(T == bool)) {
+      return boolean;
+    }
```
```d
else static assert(false, "Can't get " ~ T.stringof ~ " from a DBusAny, can only get integral types, double, string, ObjectPath, bool, DBusAny[] and ubyte[]");
```

thaven replied:
That assert would never trigger because the static ifs list all types allowed by the constraint. Still it makes some sense to add it, though with a different message, just because reaching the `else` would indicate a ddbus bug.

WebFreak001 replied:
oh yeah you are right, didn't pay attention to that, just saw an unterminated `else static if` block. But I think there should be a static assert false without error message in case we add a new type to BasicTypes and forget to update this (like SIGNATURE or UNIX_FD)

WebFreak001 commented on source/ddbus/thin.d:
```diff
@@ -264,6 +267,66 @@ struct DBusAny {
+
+    static if (isIntegral!T) {
+      enum memberName =
+        (isUnsigned!T ? "uint" : "int") ~ (T.sizeof << 3).to!string;
```
this should be `* 8`, it's more readable. The compiler should optimize such an expression anyway. Plus this is at compile time so there won't be any runtime cost anyway

WebFreak001 commented on source/ddbus/thin.d:
```diff
@@ -387,6 +468,15 @@ unittest {
     assertEqual(DBusAny(value), b);
+
+    static if(is(T == Variant!R, R)) {
+      static if (__traits(compiles, b.get!R))
```
you are mixing spaces after `if` here

WebFreak001 reviewed (changes requested):
Looks good, just add that static assert error message there for better clarity and some minor doc & code style changes

thaven committed:
Apply requested changes

- Adapt code syle to rest of project
- Added static assert in `get` overload for basic types
- Modified doc comment for `get` to include a `Returns:` section
- Rewrote shift by 3 to multiply by 8

thaven committed:
Also extend the doc comment of `typeSig`
